articles/graph-data-structure-python-part1/index.md

```diff
@@ -0,0 +1,256 @@
+
+##### Weighted Graphs
+
+A weighted graph has a value associated with every edge. The value may represent quantities like cost, distance, time, etc.., depending on the graph. An edge of a weighted graph is represented as, `(u, v, w)`.
+
+- `u` -> Source vertex
+- `v` -> Destination vertex
+- `w` -> Weight associated to go from u to v.
+
+![Weighted Graphs](/engineering-education/graph-data-structure-python-part1/weighted-graph.png)
+
+_Figure: Weighted Graph ([Source](https://www.clipartmax.com/middle/m2i8Z5i8G6b1b1b1_directed-weighted-ex1-directed-weighted-graph/))_
+
+##### Unweighted Graphs
+
+An unweighted graph does not have a value associated with every edge. An edge of an unweighted graph is represented as, `(u, v)`.
+
+- `u` -> Source vertex
+- `v` -> Destination vertex
+
+![Unweighted Graphs](/engineering-education/graph-data-structure-python-part1/unweighted-graph.png)
+
+_Figure: Unweighted Graph ([Source](https://www.researchgate.net/figure/Unweighted-Undirected-Graph_fig3_327864836))_
+
+#### Special Graphs
+
+##### Trees
+
+An undirected graph with zero cycles is called a tree. A cycle in a graph is a sequence with the first and last vertices in the sequence repeating.
+
+It has X vertices and X-1 edges.
+
+![Tree - Graphs](/engineering-education/graph-data-structure-python-part1/tree-graph.png)
+
+*Figure: Tree ([Source](https://en.wikipedia.org/wiki/Tree_(graph_theory)))*
+
+##### Rooted Tree
+
+A rooted tree is a tree which has a designated root node. If edges point away from the root, it is called an **arborescence / out-tree**. If edges point towards the root, it is called an **anti-arborescence / in-tree**.
+
+![Rooted Tree](/engineering-education/graph-data-structure-python-part1/rooted-trees.png)
+
+_Figure: Rooted Tree_
+
+##### Directed Acyclic Graphs
+
+Directed Acyclic Graphs or DAGs are directed graphs with no cycles. They represent structures with dependencies. All arborescences are DAGs, but not all DAGs are arborescences.
+
+![Directed Acyclic Graph](/engineering-education/graph-data-structure-python-part1/dag.png)
+
+_Figure: Directed Acyclic Graph ([Source](https://en.wikipedia.org/wiki/Directed_acyclic_graph))_
+
+##### Complete Graphs
+
+Complete graphs have a unique edge between every pair of vertices. A complete graph `n` vertices has `(n*(n-1)) / 2` edges and is represented by Kn.
+
+![Complete Graph](/engineering-education/graph-data-structure-python-part1/complete-graph.png)
+
+_Figure: Complete Graph ([Source](https://www.geeksforgeeks.org/mathematics-graph-theory-basics/))_
+
+### Representing Graphs
+
+There are multiple ways of using data structures to represent a graph. The three most common ways are:
+
+- [Adjacency Matrix](#adjacency-matrix)
+- [Adjacency List](#adjacency-list)
+- [Edge List](#edge-list)
+
+#### Adjacency Matrix
+
+An Adjacency Matrix is a very simple way to represent a graph. The element `**A[i][j]**` represents the cost of moving from vertex `i` to vertex `j`.
```

Comment:
Please review and correct your statements and code example about adjacency matrices. Look at https://www.programiz.com/dsa/graph-adjacency-matrix for information about them.

Reply:
@Manley12 This statement I included was right. In a weighted graph, A[i][j] represents the cost to move from vertex i to vertex j
In an unweighted graph if A[i][j] = 0, there exists no path from vertex i to j. If A[i][j] = 1, there exists a path from vertex i to j.

Reply:
Thank you for pointing that out!